### ISSUE TITLE
RFE: change indication/mail rule

### ISSUE THREAD

**Issue by zenfish** [apparition47/MailTrackerBlocker#14]
First - thanks, this is awesome.

Secondly - a small RFE... the little X that shows things is really indistinct... having even an option to make it pop more (say, changing color to red, yellow, whatever) would be really useful.

Even more greatness would be the ability to run a rule (as in preferences->rules) that would allow basically arbitrary actions (color, junk it, whatever)... I haven't looked into your code to see the magic under the hood, so not sure how possible that is (or even changing the color above, vis a vis mail APIs and such.)

Anyway, thanks again!

**Comment by zenfish** [comment]
Just reread the description... putting in an X-header would be awesome, since you already strip content out.

**Comment by apparition47** [comment]
Thanks for your input!


>a small RFE... the little X that shows things is really indistinct

The `ⓧ` icon was intended to be quite subtle as to not look out of place with the rest of the Mail app using blues and greys (see #2 CC: @bitmanic). As an aside, I'm thinking about change the icon to maybe `👁️` in line with similar extensions I've seen.

>the ability to run a rule (as in preferences->rules) that would allow basically arbitrary actions (color, junk it, whatever)

What percentage of your emails don't show the blue or grey `ⓧ` icon? The issue with email nowadays is that nearly every single email includes some sort of tracker. I imagine the end result being that your rules would trigger and the majority of your emails end up coloured or in the junk. So I'm not sure if this would be a helpful feature.

>putting in an X-header would be awesome, since you already strip content out.

This makes sense. I'll see if I can get around to this, thanks!

**Comment by zenfish** [comment]
X-header would rock, thanks for the consideration.

As for what percent... well, what percent from friends (hopefully not much :)), from work (again!), from random folks... well, who knows.  From retailers and such, sure.  But to me it's not about the high-percent that have them, which is somewhat akin to asking what % is spam - 99%+?  But Of the ones that manage to get through your defenses... I'd personally love to be able to mark or flag any and all that have this if I wish.  It'd be especially interesting/pernicious if a sender that you thought you trusted (friend or biz or w/e) was flagged, that'd be a big deal.  For the amazons/etc of the world, sure, that's noise.

I've been aware of this forever, but am interested in seeing how pervasive it is in places where I wouldn't normally expect it... for instance, the alert from github on your response - "MailTrackerBlocker blocked a tracker from GitHub."  What?!?  I may be naive, but I didn't expect it.  And really, I didn't even see the alert until I went back while writing this.  A fine line between informative and invisible, or something like that.

You say (probably correctly) that most mail does have them... but it does seem that having control of how to present that to the end consumer is critical to better understanding.  Yes, you don't want to simply wash out everything in a big red smear.  

But... wait a second... don't tell me.... Brutus, et tu?

**Comment by apparition47** [comment]
Definitely. Trackers are definitely way more pervasive than I think a lot of people realize and moreover, just as easy for for everyone and their neighbour to implement. [**@dhh** (who inspired this project) has been in the spotlight a lot over this very issue](https://twitter.com/dhh/status/1362360848430358534?s=20). I've had coworkers send me emails with trackers too. It's unbelievable. And so MailTrackerBlocker operates on this very premise of trackers being so prevalent that as you said, after the 10th email, any sort of flashy or overbearing UI indicator would end up being noise.

But I've been considering minor changes to the indicator. Perhaps something along the lines of `Ugly Email` where they have a `👁️` icon on the Conversation List if a tracker is detected.

![](https://addons.cdn.mozilla.net/user-media/previews/thumbs/203/203626.png?modified=1543521360)

Your other points noted though for having options. Maybe I'll see if I can hook into the `Rules` with an extra condition for something like "Message has tracker". A summary/stats page like you said might be interesting too.

![](https://thesweetsetup.com/wp-content/uploads/2019/08/mail-rule-conditions.jpeg)
)

**Comment by bitmanic** [comment]
Yeah - great ideas, @zenfish! 

@apparition47: if you would like me to generate any new assets related to this work, please let me know. I'm definitely interested in mocking up some summary/stat page concepts if and when you're ready to tackle that feature.

**Comment by caponecicero** [comment]
I'll second/third/fourth/whatever the request for an X-header.  I have filters that run in the delivery pipeline that do this.  It makes writing rules (without mucking with Mail to add new rules conditions).

**Comment by caponecicero** [comment]
> I'll second/third/fourth/whatever the request for an X-header. I have filters that run in the delivery pipeline that do this (I operate my own mail server). It makes writing rules (without mucking with Mail to add new rules conditions) really easy.
> 

Please see [leggett/simplify-trackers/ruby](https://github.com/leggett/simplify-trackers/tree/main/ruby).  I implemented a simple ruby filter that does exactly this.  See the [README.md](https://github.com/leggett/simplify-trackers/tree/main/ruby/README.md).

**Comment by apparition47** [comment]
>rules engine be packaged as a separate component

@caponecicero I like this idea. For your Ruby filter program, what does the email input/format look like? e.g. something standard like the `eml` files that come out of Mail with the headers and body?

**Comment by caponecicero** [comment]
You write:

> @caponecicero I like this idea. For your Ruby filter program, what does the email input/format look like? e.g. something standard like the `eml` files that come out of Mail with the headers and body?

That's exactly right...  The input/output format is a mail message.  I put it into my delivery pipeline today to read the email from STDN, modify it in-place if necessary, and write the (possibly-modified) email to STDOUT (that's what `sieve` requires of filters).  If the message has trackers, they are all removed and a header called `X-Blocked-Trackers` is added to the message with a list of the names of the trackers that were blocked.
